<launch>

    <node pkg="tcp_receive_pkg" type="tcp_receive_node" name="tcp_receive_node" />
    
    
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
    <!-- 坐标系 -->
    <param name="robot_base_frame" value="base_link"/>
    <param name="odom_frame" value="odom"/>
    <param name="map_frame" value="map"/>

    <!-- 粒子滤波 -->
    <param name="particles" value="60"/>        <!-- 稳定性更好 -->
    <param name="resampleThreshold" value="0.3"/> <!-- 减少频繁重采样 -->

    <!-- 更新策略 -->
    <param name="linearUpdate" value="0.25"/>   <!-- 移动 25cm 更新 -->
    <param name="angularUpdate" value="0.2"/>   <!-- 旋转约 11° 更新 -->
    <param name="temporalUpdate" value="2.0"/>  <!-- 每2秒触发一次 -->
    <param name="map_update_interval" value="1.0"/>

    <!-- 雷达参数 -->
    <param name="maxRange" value="4.5"/>        <!-- 雷达最大距离 -->
    <param name="maxUrange" value="4.0"/>       <!-- 有效距离 -->
    <param name="minimumScore" value="200"/>    <!-- 匹配得分阈值 -->

    <!-- 匹配精度 -->
    <param name="sigma" value="0.05"/>          <!-- 噪声模型稍紧 -->
    <param name="lsigma" value="0.05"/>         <!-- 激光匹配误差 -->
    <param name="kernelSize" value="3"/>        <!-- 匹配窗口扩大 -->
    <param name="srr" value="0.1"/>             <!-- 里程计平移噪声 -->
    <param name="srt" value="0.2"/>
    <param name="str" value="0.1"/>
    <param name="stt" value="0.2"/>

    <!-- 地图参数 -->
    <param name="xmin" value="-3.0"/>
    <param name="ymin" value="-3.0"/>
    <param name="xmax" value="3.0"/>
    <param name="ymax" value="3.0"/>
    <param name="delta" value="0.03"/>          <!-- 分辨率 3cm -->
    <param name="occ_thresh" value="0.3"/>      <!-- 占据阈值提高一点 -->
    <param name="maxMapSizeX" value="2000"/>
    <param name="maxMapSizeY" value="2000"/>

    <remap from="scan" to="/scan"/>
    <remap from="odom" to="/odom"/>
</node>


    <node pkg="slam_nav_pkg" type="global_costmap_node" name="global_costmap_node" >
    <param name="inflation_radius_m" value="0.27"/>
    <param name="map_topic" value="/map"/>
    <param name="costmap_topic" value="/global_costmap"/>
    <param name="threshold" value="50"/>
    <param name="max_cost" value="100"/>
    <param name="min_cost" value="1"/>
    </node>
    
    <node pkg="slam_nav_pkg" type="local_costmap_node" name="local_costmap_node" >
    <param name="robot_base_frame" value="base_link"/>
    <param name="map_frame" value="odom"/>
    <param name="costmap_topic" value="/local_costmap"/>
    <param name="radius_m" value="2.0"/>
    <param name="resolution" value="0.05"/>
    <param name="inflation_radius" value="0.14"/>
    </node>

    <node pkg="slam_nav_pkg" type="global_planner_node" name="global_planner" output="screen"/>

    <node pkg="rviz" type="rviz" name="rviz" />

</launch>